First take at a simple Schema, 4

lasseebert opened the pull request:
Ref #4 

A simple schema. It can be build from a list of rules.

Example of a simple schema that only validates that the value of `:name` is filled:

```elixir
schema = Schema.build([Rule.built_in("value", :name, Predicate.built_in("filled?"))])

params = %{name: ""}
result = schema.val.(params)

assert result.valid? == false
assert result.data == %{name: ""}
assert result.errors == %{name: ["must be filled"]}
```

The schema structure looks like this:

```elixir
%Validation.Schema{
  meta: [
    rules: [
      %Validation.Rule{
        meta: [
          type: "value",
          key: :name,
          predicate: %Validation.Predicate{
            meta: [type: "basic", name: "filled?", message: "must be filled"],
            val: #Function<1.42827446/1 in Validation.Predicate.build_basic/3>
          }
        ],
        val: #Function<2.100964708/1 in Validation.Rule.built_in/3>
      }
    ]
  ],
  val: #Function<1.96799137/1 in Validation.Schema.build/1>
}
```

Notice how each "piece" of the schema has it's own `val` function. This is the compiled validator function. The schema `val` uses the rule `val`, which in turn uses the predicate `val`.

lasseebert committed:
First take at a simple Schema, 4

lasseebert commented:
I imagine that it would be easy to build a nice DSL on top of this so one could transform the not-so-nice

``` elixir
schema = Schema.build([Rule.built_in("value", :name, Predicate.built_in("filled?"))])
```

into something like

``` elixir
schema = Validation.schema do
  value(:name, filled?)
end
```

lasseebert commented:
Next I will focus on being able to implement realistic (still not nested) schemas that uses required/optional rules in conjunction with more complex predicates. This will probably just be accomplished by adding a few built-in predicates and rules.

I will also have a look at how to white-list params and/or give an error if a key is given for which there is no rule. IMO we should be very explicit about this since a schema that validates form values might want to white-list and a schema that validates params in an API endpoint might want to return an error if an unknown param key is given.

certainty commented on lib/validation/schema.ex:
```diff
@@ -0,0 +1,25 @@
+    val = fn params ->
+      result = %Result{data: params}
+      rules
+      |> Enum.reduce(result, fn rule, result ->
```
I think this could be simplified a bit. I would suggest to alter the interface of a `Validation.Rule` a tiny bit.
I propose to add the following: 

``` elixir
 def apply(%__MODULE__{ val: val },  result), do: val.(result)
```

This had the benefit of hiding the detail, that rules are implemented as compiled functions and provides a more generic interface that just depends on the module's struct.

With this function the code above can become:

``` elixir
alias Validation.Rule
Enum.reduce(rules, result, &Rule.apply/2) 
```

This suggestion has the downside of having roughly one additional reduction due to the indirection of `apply/2`.
I think though that the advantage could justify it. (As an aside I think the apply interface could also be added to `Validation.Schema`)

What do you think?

lasseebert replied:
I think you are absolutely right about this.
Making Rule the module that knows how to apply a rule to a result seems like the only reasonable thing to do :)
Adding an `apply` to Schema is good for consistency.

Will fix,

certainty commented on test/schema_test.exs:
```diff
@@ -0,0 +1,20 @@
+  alias Validation.Rule
+  alias Validation.Schema
+
+  test "using a simple schema" do
```
I think a test that verifies that valid data isn't treated as invalid by the schema should be added here.
What's your thought?

lasseebert replied:
Agreed. Will fix.

lasseebert committed:
Added schema test for valid params, 4

lasseebert committed:
Refactor: Moved rule apply logic to Rule, 4

lasseebert committed:
Added apply function to Schema, 4

lasseebert committed:
Refactor: Moved predicate apply logic to Predicate, 4

lasseebert commented:
@certainty: Great feedback :) :+1:
I added an `apply` function to Predicate, Rule and Schema. Also added a schema test for valid params as you suggested.

certainty commented:
This looks very good. Keep up the awesome work!

lasseebert commented:
Thanks! :)
